src/Microsoft.Extensions.Configuration.AzureAppConfiguration/ConfigurationClients/FailOverSupportedConfigurationClient.cs

```diff
@@ -0,0 +1,256 @@
+                                                                                .AsPages(continuationToken: continuationToken)
+                                                                                .GetAsyncEnumerator();
+
+                Page<ConfigurationSetting> page = enumerator != null && await enumerator.MoveNextAsync() ? enumerator.Current : null;
+                return page;
+            }
+
+            do
+            {
+                settingsPage = await ExecuteWithFailOverPolicyAsync(func, cancellationToken);
+                if (settingsPage != null)
+                {
+                    result.AddRange(settingsPage.Values);
+                    continuationToken = settingsPage.ContinuationToken;
+                }
+            }
+            while (settingsPage != null && settingsPage.ContinuationToken != null);
+
+            return result;
+        }
+
+        public void UpdateSyncToken(Uri endpoint, string syncToken)
+        {
+            this._clients.Single(clientAndState => clientAndState.Endpoint.Host.Equals(endpoint.Host)).Client.UpdateSyncToken(syncToken);
+        }
+
+        private async Task<T> ExecuteWithFailOverPolicyAsync<T>(Func<ConfigurationClient, Task<T>> funcToExecute, CancellationToken cancellationToken = default)
+        {
+            var tasks = new List<Task>();
+            Exception lastException = null;
+            IEnumerable<ConfigurationClientState> clients = GetPrioritizedConfigurationClientList();
+
+            using (var attemptsCancellation = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken))
+            {
+                foreach (ConfigurationClientState client in clients)
+                {
+                    var parallelAwait = Task.Delay(this._parallelRetryInterval, attemptsCancellation.Token);
+                    Task<T> funcTask = funcToExecute(client.Client);
+                    tasks.Add(funcTask);
+                    tasks.Add(parallelAwait);
+
+                    var completedTask = await Task.WhenAny(tasks);
+                    if (!completedTask.Equals(parallelAwait))
+                    {
+                        if (completedTask.Status == TaskStatus.RanToCompletion)
+                        {
+                            client.BackoffEndTime = DateTimeOffset.UtcNow;
+                            client.FailedAttempts = 0;
+
+                            // Safe because the task is completed.
+                            return funcTask.Result;
+                        }
+                        else if (completedTask.Status == TaskStatus.Faulted)
+                        {
+                            tasks.Remove(completedTask);
+                            tasks.Remove(parallelAwait);
+
+                            if (IsRetryableException(completedTask.Exception))
+                            {
+                                lastException = completedTask.Exception;
+                                client.FailedAttempts++;
```

Comment:
This can be accessed by multiple threads. Need to use Interlocked.Increment.

Reply:
`FailedAttempts` is a property which can't be passed to `Interlocked.Increment` as a `ref` value.
I can possibly use
```csharp
lock(client)
{
    client.FailedAttempts++;
}
```

Reply:
It's a private type, we could use field instead of property to circumvent that.